MLP tutorials update

lukasgd opened the pull request:
This updates the existing MLP tutorials, to be merged with ML software in a next step. Also for the future, a Megatron-LM tutorial with recommended settings is left.

RMeli commented on docs/access/jupyterlab.md:
```diff
@@ -199,14 +199,14 @@ Examples of notebooks with `ipcmagic` can be found [here](https://github.com/
 
 !!! warning "torchrun with virtual environments"
-    When using a virtual environment on top of a base image with Pytorch, always replace `torchrun` with `python -m torch.distributed.run` to pick up the correct Python environment. Otherwise, the system Python environment will be used and virtual environment packages not available. If not using virtual environments such as with a self-contained Pytorch container, `torchrun` is equivalent to `pyt ...
+    When using a virtual environment on top of a base image with PyTorch, always replace `torchrun` with `python -m torch.distributed.run` to pick up the correct Python environment. Otherwise, the system Python environment will be used and virtual environment packages not available. If not using virtual environments such as with a self-contained PyTorch container, `torchrun` is equivalent to `pyt ...
```
```suggestion
    When using a virtual environment on top of a base image with PyTorch, always replace `torchrun` with `python -m torch.distributed.run` to pick up the correct Python environment. Otherwise, the system Python environment will be used and virtual environment packages will not available. If not using virtual environments such as with a self-contained PyTorch container, `torchrun` is equivalent to `python -m torch.distributed.run`.
```

RMeli commented on docs/access/jupyterlab.md:
```diff
@@ -216,19 +216,20 @@ Alternatively to using these launchers, it is also possible to use Slurm to obta
 
 !!! warning "Multi-GPU training from a shared Jupyter process"
-    Running multi-GPU training workloads directly from the shared Jupyter process is generally not recommended due to potential inefficiencies and correctness issues (cf. the [Pytorch docs](https://docs.pytorch.org/docs/stable/notes/cuda.html#use-nn-parallel-distributeddataparallel-instead-of-multiprocessing-or-nn-dataparallel)). However, if you need it to e.g. reproduce existing results, it is p ...
+    Running multi-GPU training workloads directly from the shared Jupyter process is generally not recommended due to potential inefficiencies and correctness issues (cf. the [PyTorch docs](https://docs.pytorch.org/docs/stable/notes/cuda.html#use-nn-parallel-distributeddataparallel-instead-of-multiprocessing-or-nn-dataparallel)). However, if you need it to e.g. reproduce existing results, it is p ...
```
```suggestion
    Running multi-GPU training workloads directly from the shared Jupyter process is generally not recommended due to potential inefficiencies and correctness issues (cf. the [PyTorch docs](https://docs.pytorch.org/docs/stable/notes/cuda.html#use-nn-parallel-distributeddataparallel-instead-of-multiprocessing-or-nn-dataparallel)). However, if you need it to e.g. reproduce existing results, it is possible to do so with utilities like `accelerate`'s `notebook_launcher` or [`transformers`](https://github.com/huggingface/transformers)' `Trainer` class. When using these in containers, you will currently need to unset the environment variables `RANK` and `LOCAL_RANK` by adding the following in a cell at the top of the notebook:
```

RMeli commented on docs/guides/mlp_tutorials/index.md:
```diff
@@ -1,11 +1,10 @@
-3. [Nanotron Training][ref-mlp-llm-nanotron-tutorial]
+These tutorials gradually introduce key concepts of the Machine Learning Platform. A particular focus is on the [Container Engine][ref-container-engine] for managing the runtime environment.
 
+In a [first tutorial][ref-mlp-llm-inference-tutorial], you will learn how to run an inference with an LLM on a single node using a container from the NVIDIA GPU Cloud (NGC). Concepts such as container environment description, layering a thin virtual environment on top of the container image and job launching and monitoring will be introduced.
```
```suggestion
In a [first tutorial][ref-mlp-llm-inference-tutorial], you will learn how to run inference with a LLM on a single node using a container from the NVIDIA GPU Cloud (NGC). Concepts such as container environment description, layering a thin virtual environment on top of the container image, and job launching and monitoring will be introduced.
```

RMeli commented on docs/guides/mlp_tutorials/index.md:
```diff
@@ -1,11 +1,10 @@
 
+In a [first tutorial][ref-mlp-llm-inference-tutorial], you will learn how to run an inference with an LLM on a single node using a container from the NVIDIA GPU Cloud (NGC). Concepts such as container environment description, layering a thin virtual environment on top of the container image and job launching and monitoring will be introduced.
 
+Building on the first tutorial, in the [second tutorial][ref-mlp-llm-fine-tuning-tutorial] you will learn how to train (fine-tune) an LLM on multiple GPUs on a single node. For this purpose, you will use HuggingFace's `accelerate` and see best practices for dataset management.
```
```suggestion
Building on the first tutorial, in the [second tutorial][ref-mlp-llm-fine-tuning-tutorial] you will learn how to train (fine-tune) a LLM on multiple GPUs on a single node. For this purpose, you will use HuggingFace's `accelerate` and see best practices for dataset management.
```

RMeli commented on docs/guides/mlp_tutorials/index.md:
```diff
@@ -1,11 +1,10 @@
 
+Building on the first tutorial, in the [second tutorial][ref-mlp-llm-fine-tuning-tutorial] you will learn how to train (fine-tune) an LLM on multiple GPUs on a single node. For this purpose, you will use HuggingFace's `accelerate` and see best practices for dataset management.
 
+In the [third tutorial][ref-mlp-llm-nanotron-tutorial], you will apply the techniques from the previous tutorials to enable distributed (pre-)training of a model `nanotron` on multiple nodes. In particular, this tutorial makes use of model-parallelism and introduces the usage of `torchrun` to manage jobs on individual nodes.
```
```suggestion
In the [third tutorial][ref-mlp-llm-nanotron-tutorial], you will apply the techniques from the previous tutorials to enable distributed (pre-)training of the `nanotron` model on multiple nodes. In particular, this tutorial makes use of model-parallelism and introduces the usage of `torchrun` to manage jobs on individual nodes.
```
?

RMeli commented on docs/guides/mlp_tutorials/llm-fine-tuning.md:
````diff
@@ -8,45 +8,49 @@ This means that we take the model and train it on some new custom data to change
-$ source ./gemma-venv/bin/activate
-$ python -m pip install peft==0.11.1
+```bash
+$ cd $SCRATCH/tutorials/gemma-7b
+$ srun --environment=./ngc-pytorch-gemma-24.01.toml --pty bash
+$ source venv-gemma-24.01/bin/activate
+$ pip install peft==0.11.1
````
```suggestion
cd $SCRATCH/tutorials/gemma-7b
srun --environment=./ngc-pytorch-gemma-24.01.toml --pty bash
source venv-gemma-24.01/bin/activate
pip install peft==0.11.1
```
Since it was changed to `bash`.

RMeli commented on docs/guides/mlp_tutorials/llm-fine-tuning.md:
````diff
@@ -8,45 +8,49 @@ This means that we take the model and train it on some new custom data to change
 
-```console
+```bash
 $ git clone https://github.com/huggingface/trl -b v0.7.11
````
```suggestion
git clone https://github.com/huggingface/trl -b v0.7.11
```

RMeli commented on docs/guides/mlp_tutorials/llm-fine-tuning.md:
````diff
@@ -8,45 +8,49 @@ This means that we take the model and train it on some new custom data to change
+```bash
 $ git clone https://github.com/huggingface/trl -b v0.7.11
-$ pip install -e ./trl   # install in editable mode
+$ pip install -e ./trl # (1)!
````
```suggestion
pip install -e ./trl # (1)!
```

RMeli commented on docs/guides/mlp_tutorials/llm-fine-tuning.md:
````diff
@@ -73,38 +77,50 @@ accelerate launch --config_file trl/examples/accelerate_configs/multi_gpu.yaml \
+Make this script executable with
+
+```bash
+$ chmod u+x $SCRATCH/tutorials/gemma-7b/fine-tune-gemma.sh
````
```suggestion
chmod u+x $SCRATCH/tutorials/gemma-7b/fine-tune-gemma.sh
```

RMeli commented on docs/guides/mlp_tutorials/llm-fine-tuning.md:
````diff
@@ -115,8 +131,8 @@ Now that we've setup a fine-tuning script and a Slurm batch script, we can launc
-```console
-$ sbatch --nodes=1 fine-tune-sft.sbatch
+```bash
+$ sbatch --nodes=1 submit-fine-tune-gemma.sh
````
```suggestion
sbatch --nodes=1 submit-fine-tune-gemma.sh
```

RMeli commented on docs/guides/mlp_tutorials/llm-fine-tuning.md:
````diff
@@ -130,8 +146,8 @@ input_text = "What are the 5 tallest mountains in the Swiss Alps?"
-```console
-$ sbatch ./gemma-inference.sbatch
+```bash
+$ sbatch submit-gemma-inference.sh
````
```suggestion
sbatch submit-gemma-inference.sh
```

RMeli commented on docs/guides/mlp_tutorials/llm-inference.md:
```diff
@@ -5,54 +5,69 @@
 However, chances are that we will need some additional libraries or software.
-For this reason, we need to use some docker commands to build a container on top of what is provided by Nvidia.
-To do this, we create a new directory for building containers in our home directory and set up a [Dockerfile](https://docs.docker.com/reference/dockerfile/):
+For this reason, we need to use some docker commands to build on top of what is provided by Nvidia.
```
```suggestion
For this reason, we need to use some podman commands to build on top of what is provided by Nvidia.
```
?

RMeli commented on docs/guides/mlp_tutorials/llm-inference.md:
````diff
@@ -5,54 +5,69 @@
 
-```console
+```bash
 $ cd $SCRATCH
````
```suggestion
cd $SCRATCH
```

RMeli commented on docs/guides/mlp_tutorials/llm-inference.md:
````diff
@@ -5,54 +5,69 @@
+```bash
 $ cd $SCRATCH
-$ mkdir pytorch-24.01-py3-venv && cd pytorch-24.01-py3-venv
+$ mkdir -p tutorials/gemma-7b
+$ cd tutorials/gemma-7b
````
```suggestion
mkdir -p tutorials/gemma-7b
cd tutorials/gemma-7b
```

RMeli commented on docs/guides/mlp_tutorials/llm-inference.md:
````diff
@@ -62,77 +77,117 @@ This step is straightforward, just make the file `$HOME/.config/containers/stora
+Before building the container image, we create a dedicated directory to keep track of all images used with the CE. Since container images are large files and the filesystem is a shared resource, we need to apply [best practices for LUSTRE][ref-guides-storage-lustre] so they are properly distributed across storage nodes.
+
+```bash title="Container image directory with recommended LUSTRE settings"
+$ mkdir -p $SCRATCH/ce-images
+$ lfs setstripe -E 4M -c 1 -E 64M -c 4 -E -1 -c -1 -S 4M $SCRATCH/ce-images # (1)!
````
```suggestion
mkdir -p $SCRATCH/ce-images
lfs setstripe -E 4M -c 1 -E 64M -c 4 -E -1 -c -1 -S 4M $SCRATCH/ce-images # (1)!
```

RMeli reviewed (commented):
Mostly nitpicks (they can be accepted/rejected from the GitHub UI), but otherwise LGTM.

I think it is good to switch from `console` to `bash` in most cases (`bash` should not have the shell prompt `$`), but `console` should be preferred when there is a mix of commands and output (commands need the prompt `$`) or when it is important to show the prompt modification (as in virtual environments).

lukasgd commented:
As discussed with @RMeli, I've reverted the bash blocks to console and made explicit where a command is run by adding the hostname and possibly virtual environment to the shell prompt. I've also added a paragraph about the login node policy.
